fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,364 @@
+
+use crate::fvm::state::FvmStateParams;
+use crate::fvm::store::ReadOnlyBlockstore;
+use anyhow::anyhow;
+use cid::multihash::{Code, MultihashDigest};
+use cid::Cid;
+use futures_core::Stream;
+use fvm::state_tree::StateTree;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::{load_car_unchecked, CarHeader};
+use fvm_ipld_encoding::{from_slice, CborStore, DAG_CBOR};
+use libipld::Ipld;
+use serde::{Deserialize, Serialize};
+use std::collections::VecDeque;
+use std::path::Path;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio_stream::StreamExt;
+use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+pub type BlockHeight = u64;
+
+/// Taking snapshot of the current blockchain state
+pub enum Snapshot<BS> {
+    V1(V1Snapshot<BS>),
+}
+
+/// Contains the overall metadata for the snapshot
+#[derive(Serialize, Deserialize)]
+struct SnapshotMetadata {
+    version: u8,
+    data_root_cid: Cid,
+}
+
+/// The streamer that streams the snapshot into (Cid, Vec<u8>) for car file.
+type SnapshotStreamer = Box<dyn Send + Unpin + Stream<Item = (Cid, Vec<u8>)>>;
+
+impl<BS> Snapshot<BS>
+where
+    BS: Blockstore + Clone + 'static + Send,
+{
+    pub fn new(
+        store: BS,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        Ok(Self::V1(V1Snapshot::new(
+            store,
+            state_params,
+            block_height,
+        )?))
+    }
+
+    /// Read the snapshot from file and load all the data into the store
+    pub async fn read_car(path: impl AsRef<Path>, store: BS) -> anyhow::Result<Self> {
+        let file = tokio::fs::File::open(path).await?;
+
+        let roots = load_car_unchecked(&store, file.compat()).await?;
+        if roots.len() != 1 {
+            return Err(anyhow!("invalid snapshot, should have 1 root cid"));
+        }
+
+        let metadata_cid = roots[0];
+        let metadata = if let Some(metadata) = store.get_cbor::<SnapshotMetadata>(&metadata_cid)? {
+            metadata
+        } else {
+            return Err(anyhow!("invalid snapshot, metadata not found"));
+        };
+
+        match metadata.version {
+            1 => Ok(Self::V1(V1Snapshot::from_root(
+                store,
+                metadata.data_root_cid,
+            )?)),
+            v => Err(anyhow!("unknown snapshot version: {v}")),
+        }
+    }
+
+    /// Write the snapshot to car file
+    pub async fn write_car(self, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        let file = tokio::fs::File::create(path).await?;
+
+        // derive the metadata for the car file, so that the snapshot version can be recorded.
+        let (metadata, snapshot_streamer) = self.into_streamer()?;
+        let (metadata_cid, metadata_bytes) = derive_cid(&metadata)?;
+
+        // create the target car header with the metadata cid as the only root
+        let car = CarHeader::new(vec![metadata_cid], 1);
```

Comment:
So the root of our snapshot is the metadata (i.e. the only object identified in the header) and from the metadata we get the root cid of the actual snapshot, right? I think it may be useful to document this, one may have also expected the header to include two cids, one for the metadata and one for the snapshot data.

(That being said, I think I like our approach more, but let's document so it's clear what we are doing)

Reply:
Comments added!